src/runtime.cpp

```diff
@@ -16,186 +16,227 @@
+    SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_ES);
+    SDL_SetHint(SDL_HINT_EMSCRIPTEN_KEYBOARD_ELEMENT, "#canvas");
+#endif
 
-    SDL_Event mapEvent(SDL_Event evt, double scaleFactor)
+    // create window
+    DPI::ScaleFactor = get_scale_factor();
+    int winWidth = w * DPI::ScaleFactor;
+    int winHeight = h * DPI::ScaleFactor;
+    SDL_Window* window =
+#ifndef EMSCRIPTEN
+      SDL_CreateWindow(title.c_str(),
+#else
+      SDL_CreateWindow(nullptr,
+#endif
+                       SDL_WINDOWPOS_CENTERED,
+                       SDL_WINDOWPOS_CENTERED,
+                       winWidth,
+                       winHeight,
+                       SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE | SDL_WINDOW_ALLOW_HIGHDPI);
+    if (!window)
     {
-      if (evt.type == SDL_MOUSEMOTION)
-      {
-        double x1 = evt.motion.x;
-        double y1 = evt.motion.y;
-        double x0 = x1 - evt.motion.xrel;
-        double y0 = y1 - evt.motion.yrel;
-        x1 = (x1 / scaleFactor - offset.x) / zoom;
-        y1 = (y1 / scaleFactor - offset.y) / zoom;
-        x0 = (x0 / scaleFactor - offset.x) / zoom;
-        y0 = (y0 / scaleFactor - offset.y) / zoom;
-        evt.motion.x = FLOAT2SINT(x1);
-        evt.motion.y = FLOAT2SINT(y1);
-        evt.motion.xrel = FLOAT2SINT(x1 - x0);
-        evt.motion.yrel = FLOAT2SINT(y1 - y0);
-      }
-      return evt;
+      handle_sdl_error();
+      return false;
+    }
+    m_width = w;
+    m_height = h;
+    m_sdlState.window = window;
+    // create context
+    SDL_GLContext glContext = SDL_GL_CreateContext(window);
+    if (!glContext)
+    {
+      handle_sdl_error();
+      return false;
     }
-  };
-  Zoomer m_zoomer;
 
-public:
-  inline void setOnFrameOnce(std::function<void()>&& cb)
-  {
-    Scheduler::setOnFrameOnce(std::move(cb));
-  }
+    m_sdlState.glContext = glContext;
 
-  inline void startRunMode()
-  {
-    EntityManager::setInteractionMode(EntityManager::InteractionMode::RUN);
-    InputManager::setMouseCursor(MouseEntity::CursorType::NORMAL);
-    if (auto container = EntityManager::getEntities())
+    // switch to this context
+    if (makeContextCurrent() == false)
     {
-      container->setRunModeInteractions();
+      return false;
     }
-  }
+    INFO("GL_VENDOR: %s", glGetString(GL_VENDOR));
+    INFO("GL_RENDERER: %s", glGetString(GL_RENDERER));
+    INFO("GL_VERSION: %s", glGetString(GL_VERSION));
+    INFO("GL_SHADING_LANGUAGE_VERSION: %s", glGetString(GL_SHADING_LANGUAGE_VERSION));
 
-protected:
-  virtual void onInit() override
-  {
-    SDL_ShowCursor(SDL_DISABLE);
+    // get device pixel ratio
+    int dw = 0, dh = 0;
+    int ww = 0, wh = 0;
+    SDL_GL_GetDrawableSize(window, &dw, &dh);
+    SDL_GetWindowSize(window, &ww, &wh);
+    m_properties["viewport_w"] = dw;
+    m_properties["viewport_h"] = dh;
+    m_properties["window_w"] = ww;
+    m_properties["window_h"] = wh;
+    m_properties["app_w"] = w;
```

Comment:
Duplication with member data `m_width` and `m_height`. If we use the properties, we should turn other member data into properties as well. And properties need both getter and setter, like `getProperty` and `setProperty`. Currently only getter exists.

Or we just don't use properties if there are no actual benefits.